### ISSUE TITLE
unnecessary_operation false positive with Drop

### ISSUE THREAD

**Issue by jugglerchris** [rust-lang-nursery/rust-clippy#2061]
On the following code:

```rust
use std::rc::Rc;

struct Foo {
    _r: Rc<u32>,
}

impl Drop for Foo {
    fn drop(&mut self) {
        println!("Foo");
    }
}

fn main() {
    Foo { _r: Rc::new(1) };  // clippy: replace with "Rc::new(1)"
}
```

clippy suggests removing the `Foo{}` wrapper, which would cause `Foo::drop` not to be called:

```
warning: statement can be reduced
  --> src/main.rs:14:5
   |
14 |     Foo { _r: Rc::new(1) };
   |     ^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `Rc::new(1);`
   |
   = note: #[warn(unnecessary_operation)] on by default
```

**Comment by oli-obk** [comment]
We should not lint on drop types, except libstd ones known not to have side effects other than allocations
